src/EventStore.Core/Services/Transport/Grpc/Enumerators.AllSubscription.cs

```diff
@@ -40,95 +51,412 @@ public class AllSubscription : IAsyncEnumerator<ResolvedEvent> {
+
+				await _inner.DisposeAsync().ConfigureAwait(false);
+
+				_inner = new CatchupAllSubscription(_subscriptionId, _bus, caughtUpPosition, _resolveLinks, _user,
+					_readIndex, _cancellationToken);
+			}
+
+			private class CatchupAllSubscription : IAsyncEnumerator<ResolvedEvent> {
+				private readonly Guid _subscriptionId;
+				private readonly IPublisher _bus;
+				private readonly bool _resolveLinks;
+				private readonly IPrincipal _user;
+				private readonly CancellationTokenSource _disposedTokenSource;
+				private readonly ConcurrentQueue<ResolvedEvent> _buffer;
+				private readonly CancellationTokenRegistration _tokenRegistration;
+				private Position _nextPosition;
+				private ResolvedEvent _current;
+
+				public ResolvedEvent Current => _current;
+
+				public CatchupAllSubscription(Guid subscriptionId,
+					IPublisher bus,
+					Position position,
+					bool resolveLinks,
+					IPrincipal user,
+					IReadIndex readIndex,
+					CancellationToken cancellationToken) {
+					if (bus == null) {
+						throw new ArgumentNullException(nameof(bus));
+					}
+
+					if (readIndex == null) {
+						throw new ArgumentNullException(nameof(readIndex));
+					}
+
+					_subscriptionId = subscriptionId;
+					_bus = bus;
+					_nextPosition = position == Position.End
+						? Position.FromInt64(readIndex.LastIndexedPosition, readIndex.LastIndexedPosition)
+						: position;
+					_resolveLinks = resolveLinks;
+					_user = user;
+					_disposedTokenSource = new CancellationTokenSource();
+					_buffer = new ConcurrentQueue<ResolvedEvent>();
+					_tokenRegistration = cancellationToken.Register(_disposedTokenSource.Dispose);
+					Log.Info("Catch-up subscription {subscriptionId} to $all running...",
+						_subscriptionId);
 				}
 
-				if (_buffer.TryDequeue(out current)) {
-					_current = current;
-					return true;
+				public ValueTask DisposeAsync() {
+					_disposedTokenSource.Dispose();
+					_tokenRegistration.Dispose();
+					return default;
 				}
 
-				try {
-					await Task.Delay(100, _disposedTokenSource.Token).ConfigureAwait(false);
-				} catch (ObjectDisposedException) {
-					return false;
+				public async ValueTask<bool> MoveNextAsync() {
+					ReadLoop:
+					if (_disposedTokenSource.IsCancellationRequested) {
+						return false;
+					}
+
+					if (_buffer.TryDequeue(out var current)) {
+						_current = current;
+						return true;
+					}
+
+					var correlationId = Guid.NewGuid();
+
+					var readNextSource = new TaskCompletionSource<bool>();
+
+					var (commitPosition, preparePosition) = _nextPosition.ToInt64();
+
+					Log.Trace(
+						"Catch-up subscription {subscriptionId} to $all: reading next page starting from {nextPosition}.",
+						_subscriptionId, _nextPosition);
+
+					_bus.Publish(new ClientMessage.ReadAllEventsForward(
+						correlationId, correlationId, new CallbackEnvelope(OnMessage), commitPosition, preparePosition,
+						32, _resolveLinks, false, default, _user));
```

Comment:
`MaxCount` should be configurable on the subscription.
`RequireMaster` should probably also be configured on the client, though that could be handled as a separate PR